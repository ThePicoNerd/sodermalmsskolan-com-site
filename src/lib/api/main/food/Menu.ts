import useSWR from "swr";
import { fetchJSON } from "../fetch";
import { CollectionResponse } from "../Response";
import moment from "moment";

export interface Dish {
  title: string;
  id: string;
}

export interface Menu {
  dishes: Dish[];
  date: Date;
}

export type MenuResponse = CollectionResponse<Menu>;

export interface MenuQuery {
  /**
   * How many menus to retreive (maximum).
   */
  limit?: number;

  /**
   * Offset the starting day, counting from today.
   */
  offset?: number;
}

/**
 *
 */
export function useMenus({ limit = 10, offset = 0 }: MenuQuery) {
  return useSWR(`/api/food/menus?limit=${limit}&offset=${offset}`, async () => {
    const res = await fetchJSON<MenuResponse>("/api/food/menus");

    const menus = res.data.sort(({ date: a }, { date: b }) => {
      if (a < b) {
        return -1;
      }

      if (a > b) {
        return 1;
      }

      return 0;
    });

    const now = moment(new Date());
    const next = menus.findIndex((menu) =>
      now.isSameOrBefore(menu.date, "date")
    );
    const startIndex = next + offset;

    return menus.slice(startIndex, startIndex + limit);
  });
}
