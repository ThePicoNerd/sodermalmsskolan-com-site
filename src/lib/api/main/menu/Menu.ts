import { CollectionResponse } from "../Response";
import useSWR from "swr";
import { fetchJSON } from "../fetch";

export interface Menu {
  dishes: string[];
  timestamp: Date;
}

export type MenuResponse = CollectionResponse<Menu>;

export interface MenuQuery {
  /**
   * How many menus to retreive (maximum).
   */
  limit?: number;

  /**
   * Offset the starting day, counting from today.
   */
  offset?: number;
}

/**
 *
 */
export function useMenus({ limit = 10, offset = 0 }: MenuQuery) {
  return useSWR(
    `/api/menus?limit=${limit}&offset=${offset}`,
    async (url: string) => {
      const res = await fetchJSON<MenuResponse>(url);
      return res.data;
    }
  );
}
