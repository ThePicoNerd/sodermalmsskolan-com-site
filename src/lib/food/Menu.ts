export type Dishes = string[];

export default interface Menu {
  date: string;
  dishes: Dishes;
}
