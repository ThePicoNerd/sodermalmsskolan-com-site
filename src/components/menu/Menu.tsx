import dayjs from "dayjs";
import React, { FunctionComponent } from "react";
import { useLang } from "../../hooks/lang";
import { ClientMenu } from "../../lib/food/structures/client/Menu";
import DishCard from "../dishes/DishCard";
import CardGrid from "../grid/CardGrid";
import { SmallHeading } from "../text/headings";

const Menu: FunctionComponent<{menu: ClientMenu}> = ({ menu }) => {
  const lang = useLang();

  return (
    <div css={{
      display: "flex",
    }}
    >
      <SmallHeading
        css={{
          writingMode: "vertical-rl",
          marginRight: "1rem",
        }}
      >
        {menu?.date ? dayjs(menu?.date).locale(lang).format("dddd DD MMMM") : "LÃ¤ser in ..."}
      </SmallHeading>
      <CardGrid css={{
        flex: 1,
      }}
      >
        {(menu?.dishes || new Array(2).fill(null))
          .map((dish, index) => <DishCard dish={dish} key={dish?.id || index} />)}
      </CardGrid>
    </div>
  );
};

export default Menu;
