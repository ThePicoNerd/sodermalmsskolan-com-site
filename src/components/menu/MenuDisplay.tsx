import { DateTime } from "luxon";
import React, { FunctionComponent } from "react";
import Menu from "../../lib/food/Menu";
import InlineSkeleton from "../skeleton/InlineSkeleton";

export interface MenuDisplayProps {
  menu: Menu;
}

/**
 * A display for a `Menu`.
 *
 * @param {React.PropsWithChildren<MenuDisplayProps>} props Props.
 *
 * @returns {React.ReactElement} The rendered menu.
 */
const MenuDisplay: FunctionComponent<MenuDisplayProps> = ({
  menu,
}) => (
  <div className="my-4">
    <h3 className="text-xl font-semibold">
      {menu?.date
        ? DateTime.fromISO(menu?.date).toLocaleString(DateTime.DATE_HUGE)
        : <InlineSkeleton width="8em" className="max-w-full" />}
    </h3>
    <ul>
      {(menu?.dishes ?? new Array(2).fill(undefined)).map((dish, i) => (
        <li key={dish ?? i} className="text-gray-500 font-medium">{dish ?? <InlineSkeleton className="max-w-full" width="40em" />}</li>
      ))}
    </ul>
  </div>
);

export default MenuDisplay;
