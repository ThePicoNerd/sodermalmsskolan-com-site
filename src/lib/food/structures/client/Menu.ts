import useSWR, { responseInterface } from "swr";
import { IMenu, Menu, MenuQuery } from "../shared/Menu";
import ky from "ky-universal";
import dayjs from "dayjs";
import { ClientDish } from "./Dish";
import { useLocale } from "../../../../hooks/locale";
import { firstLetterUpperCase } from "../../../utils/letters";
import isSameOrBefore from "dayjs/plugin/isSameOrBefore";

dayjs.extend(isSameOrBefore);

export class ClientMenu extends Menu {
  dishes: ClientDish[];

  constructor({ dishes, date }: IMenu) {
    super({ dishes, date });

    this.dishes = dishes.map((dish) => new ClientDish(dish));
  }

  public static async fetchAll(): Promise<ClientMenu[]> {
    const res = await ky.get("/api/food/menus").json<IMenu[]>();

    const menus = res
      .map((menu) => new ClientMenu(menu))
      .sort((a, b) => a.date.getTime() - b.date.getTime());

    return menus;
  }

  public static use({
    limit = 10,
    offset = 0,
  }: MenuQuery): responseInterface<ClientMenu[], unknown> {
    return useSWR(
      `/api/food/menus?limit=${limit}&offset=${offset}`,
      async () => {
        const menus = await ClientMenu.fetchAll();

        const now = dayjs(new Date());
        const next = menus.findIndex((menu) =>
          now.isSameOrBefore(menu.date, "date")
        );
        const startIndex = next + offset;

        return menus.slice(startIndex, startIndex + limit);
      }
    );
  }

  public get title(): string {
    const { locale } = useLocale();

    return firstLetterUpperCase(
      dayjs(this?.date).locale(locale).format("dddd D MMMM")
    );
  }
}
