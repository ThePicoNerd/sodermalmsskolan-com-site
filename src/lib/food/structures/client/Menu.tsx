import useSWR from "swr";
import ky from "ky-universal";
import dayjs from "dayjs";
import isSameOrBefore from "dayjs/plugin/isSameOrBefore";
import { ClientDish } from "./Dish";
import { IMenu, Menu } from "../shared/Menu";
import { UseSWRResource } from "../../../common/usable";

dayjs.extend(isSameOrBefore);

export interface UseMenuQuery {
  limit?: number;
  offset?: number;
}

export class ClientMenu extends Menu {
  dishes: ClientDish[];

  constructor({ dishes, date }: IMenu) {
    super({ dishes, date });

    this.dishes = dishes.map((dish) => new ClientDish(dish));
  }

  public static async fetchAll(): Promise<ClientMenu[]> {
    const res = await ky.get("/api/food/menus").json<IMenu[]>();

    const menus = res
      .map((menu) => new ClientMenu(menu))
      .sort((a, b) => a.date.getTime() - b.date.getTime());

    return menus;
  }
}

export const useMenu: UseSWRResource<ClientMenu[], UseMenuQuery> = ({
  limit = 10,
  offset = 0,
}) => useSWR(`/api/food/menus?limit=${limit}&offset=${offset}`, async () => {
  const menus = await ClientMenu.fetchAll();

  const now = dayjs(new Date());
  const next = menus.findIndex((menu) => now.isSameOrBefore(menu.date, "date"));
  const startIndex = next + offset;

  return menus.slice(startIndex, startIndex + limit);
});
