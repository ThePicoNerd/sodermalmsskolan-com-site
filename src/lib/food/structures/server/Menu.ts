import { ServerDish } from "./Dish";
import { IMenu, Menu } from "../shared/Menu";
import got from "got/dist/source";
import { API_ENDPOINT } from "../../constants";

export class ServerMenu extends Menu {
  dishes: ServerDish[];

  constructor({ dishes, date }: IMenu) {
    super({ dishes, date });

    this.dishes = dishes.map((dish) => new ServerDish(dish));
  }

  public static async fetchAll(): Promise<ServerMenu[]> {
    const res = await got.get(`${API_ENDPOINT}/menu`).json<IMenu[]>();

    return res.map((menu) => new ServerMenu(menu));
  }
}
