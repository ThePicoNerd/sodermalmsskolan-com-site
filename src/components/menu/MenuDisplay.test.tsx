import { render, screen } from "@testing-library/react";
import { capitalize } from "lodash";
import { DateTime } from "luxon";
import React from "react";
import Menu from "../../lib/food/Menu";
import MenuDisplay from "./MenuDisplay";

describe("<MenuDisplay /> test", () => {
  it("should render correctly", () => {
    const menu: Menu = {
      date: "2020-01-01",
      dishes: ["Fisk Björkeby", "Tacobuffé"],
    };

    render(<MenuDisplay menu={menu} />);

    const expectedHeading = capitalize(DateTime.utc(2020, 1, 1).toLocaleString(DateTime.DATE_HUGE));

    expect(screen.getByText(/jan/i).textContent).toBe(expectedHeading);
  });

  it("should return a skeleton if the menu is undefined", () => {
    const { asFragment } = render(<MenuDisplay menu={undefined} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
