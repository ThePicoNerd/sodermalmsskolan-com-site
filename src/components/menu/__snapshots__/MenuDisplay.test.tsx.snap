// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MenuDisplay /> test should return a skeleton if the menu is undefined 1`] = `
<DocumentFragment>
  <div
    class="my-4"
  >
    <h3
      class="text-xl font-semibold"
    >
      <span>
        <span
          class="skeleton max-w-full"
          style="width: 12em;"
        >
          ‌
        </span>
      </span>
    </h3>
    <ul>
      <li
        class="text-gray-500 font-medium leading-snug my-1"
      >
        <span>
          <span
            class="skeleton max-w-full"
            style="width: 40em;"
          >
            ‌
          </span>
        </span>
      </li>
      <li
        class="text-gray-500 font-medium leading-snug my-1"
      >
        <span>
          <span
            class="skeleton max-w-full"
            style="width: 40em;"
          >
            ‌
          </span>
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
